File: algorithm/code/PHP/insert-sort.php
<?php
/**
 * @Author: Marte
 * @Date:   2018-07-30 19:31:10
 * @Last Modified by:   Marte
 * @Last Modified time: 2018-07-30 20:50:49
 */

function insert_sort(array $list) : array
{
    $data = [];
    while (!empty($list)) {

        $tmp = array_shift($list);
        $len = count($data);

        if (empty($data) || $data[$len-1] < $tmp) {

            array_push($data, $tmp);
            continue;
        }

        for($i = $len - 1; $i >= 0; $i--) {

            if ($data[$i] > $tmp) {

                $data[$i+1] = $data[$i];
                continue;
            }
            break;
        }

        $data[$i+1] = $tmp;
    }

    return $data;
}

function insert_sort_two(array $data) : array
{
    $len = count($data);
    for ($i=1; $i <= $len - 1; $i++) {
        $tmp = $data[$i];
        for ($j=$i; $j > 0 ; $j--) {
            if ($data[$j-1] < $tmp) {
                break;
            }
            $data[$j] = $data[$j-1];
        }
        $data[$j] = $tmp;
    }

    return $data;
}

$data = insert_sort_two([4,3,6,2,5,1,8,10,7]);
print_r($data);

File: algorithm/code/go/insert_sort.go
package main



// 如果某一个位置前元素都是有序的
// 那么只需要将这个元素查到之前有序列表中合适的位置即可
// 减少了比较和交换次数
// 具体实现就是，元素 与 左侧元素比较 如果小于就移动左侧元素
func InsertSort(data []int) {

	for i := 1; i< len(data); i++ {
		j := i
		tmp := data[i]
		for; j > 0; j-- {
			if data[j-1] > tmp {
				data[j] = data[j-1]
			}
		}
		data[j] = tmp
	}	
}


func main() {
    data := []int{4,7,1,5,8,3,9,1,6,3,0,3,4}
    fmt.Println(data)
    InsertSort(data)
    fmt.Println(data)
}